### ISSUE TITLE
Query on Adding Sentence Positions while creating the Graph

### ISSUE THREAD

**Issue by anilbatra2185** [shrimai/Topological-Sort-for-Sentence-Ordering#5]
Hi @shrimai,

Thanks for sharing the code and  I have following query related to code:
- The preprocessed file, which is used to create the graph, already has the ground truth sentence position in the file. However, at test time I do not have such index position/value. So, I assign value to index position based on the random order of sentences. e.g. Consider correct sequence `<s1, s2, s3, s4>` which is not known to me and I have shuffled sentence order like `<s3, s1, s4, s2>`. Based on the shuffled sentence order, I assigned index position `0 -> s3, 1 -> s1, 2 -> s4, 3 -> s2` and generated the file. Finally, ran the  [topological_sort.py](https://github.com/shrimai/Topological-Sort-for-Sentence-Ordering/blob/master/topological_sort.py) after modifying the code at [line 212](https://github.com/shrimai/Topological-Sort-for-Sentence-Ordering/blob/master/topological_sort.py#L212) to use custom order i.e. `[1, 3, 0, 2]` and obtain the metrics. Now, if I feed the `GT` index position in the pre-processed data there is large variance in the metrics.

I could not understand why do we need to add the `GT` index positions at inference time, which is not generally available e.g. randomly chosen sentences from web. I would request your help and let me know if I am doing something wrong in metric evaluations.

Thanks

**Comment by shrimai** [comment]
The ground truth index positions should only be used for evaluating the model performance (on dev and test sets for which you have ground truth). If you don't know the ground truth and imagine you have deployed the system to predict the order, then you can ignore the evaluation metrics.

**Comment by anilbatra2185** [comment]
thanks @shrimai for your response. I understand your point, however, if I imagine correctly then the output order should be same for different shuffled/permutation of inputs to the system e.g. the output of two different permutations `<s3, s1, s4, s2>` and `<s2, s1, s4, s3>` should be same i.e. correct order `<s1, s2, s3, s4>`.  Thus evaluation metric should not have large variance.

Please let me know, if I mis-understand it.

Regards
Anil

**Comment by shrimai** [comment]
This is correct. For B-TSort and L-TSort techniques we split sentences of a document in different pairs and get constraints from the pairs. I suppose you are not facing high variance with these two techniques?
Basically, whatever model you are using is susceptible to random shuffling in the input? If you provide two different permutations of the input to the model then it predicts different orders for you?
Also remember that there could be multiple 'valid' topological sorts of a DAG. Here, we only consider one of them. You can also check your prediction against multiple valid orderings.
